<a href="https://colab.research.google.com/github/pallavi-sutar097/spam_mail_prediction/blob/main/DS_project_1_spam_mail_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing the dependencies

In [ ]:
import numpy as  np
import pandas as pd
from sklearn.model_selection import train_test_split #to split the data into train and test data
from sklearn.feature_extraction.text import TfidfVectorizer #to convert the text data into numerical values(feature vectors)
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

data collection and preprocessing

In [ ]:
#load data from csv file to a pandas dataframe
raw_mail_data = pd.read_csv('/content/sample_data/mail_data.csv')

In [ ]:
print(raw_mail_data)

            Category                                            Message
0                ham  Go until jurong point, crazy.. Available only ...
1                ham                      Ok lar... Joking wif u oni...
2               spam  Free entry in 2 a wkly comp to win FA Cup fina...
3                ham  U dun say so early hor... U c already then say...
4                ham  Nah I don't think he goes to usf, he lives aro...
...              ...                                                ...
5568             ham               Will ü b going to esplanade fr home?
5569             ham  Pity, * was in mood for that. So...any other s...
5570             ham  The guy did some bitching but I acted like i'd...
5571             ham                         Rofl. Its true to its name
5572  {"mode":"full"                                    isActive:false}

[5573 rows x 2 columns]


In [ ]:
#replace the null values with null string
mail_data = raw_mail_data.where((pd.notnull(raw_mail_data)),'') #'' represents empty string

In [ ]:
#print the first 5 rows of the data frame
mail_data.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
#checking the no of rows and cols in the dataframe
mail_data.shape

(5572, 2)

the dataset has 5573 mails with labels as ham and spam
now do LABEL ENCODING: spam= 0 and ham= 1


In [ ]:
#label spam mail as 0; ham mail as 1;
mail_data.loc[mail_data['Category'] =='spam','Category',]= 0
mail_data.loc[mail_data['Category'] =='ham','Category',] = 1

spam=0
ham=1

split the data and the label, just like giving the x axis and the y axis. Here the Xaxis is message and Yaxis is the label.

In [ ]:
#separating the data as text and label
X=mail_data['Message']
Y=mail_data['Category']

In [ ]:
print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: Message, Length: 5572, dtype: object


In [ ]:
print(Y)

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: Category, Length: 5572, dtype: object


To split X and Y data into training and testing data

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, random_state=3 )
#test_size : amt of data 80% is training and 20% is testing
#random state : will be splitted in the way we want

In [ ]:
print(X.shape)
print(X_train.shape)
print(X_test.shape)

(5572,)
(4457,)
(1115,)


FEATURE EXTRACTION (using Tfifdvectorisation)
To convert the text into meaningful numerical values.

Tfifdvectorisation looks into the data , tries to go through all the words into the document and according to the number of times a word is repeated it assigns some scores to it. such as words like free,discount,offers, so on... now the log reg model will link all those words and they will get some feature vectors as numerical values with label 0 and linked to the target spam.

In [ ]:
#transform the text data into feature vectors that can be used as input to the logistic regression
feature_extraction = TfidfVectorizer(min_df=1, stop_words='english', lowercase=True) #min_df : if score of a particular word is more than 1 then we can include it coz we can use those words, stop_words: words doesnt have much meaning will be removed

X_train_features = feature_extraction.fit_transform(X_train) #fit all the data into vectorisor and transform into numerical values
X_test_features = feature_extraction.transform(X_test) # not using fit coz we dont want the model to look into the test data

#convert y_train and y_test values as integers (since it is object form)
Y_train = Y_train.astype(int)
Y_test = Y_test.astype(int)

In [ ]:
print(X_train)

3075                  Don know. I did't msg him recently.
1787    Do you know why god created gap between your f...
1614                         Thnx dude. u guys out 2nite?
4304                                      Yup i'm free...
3266    44 7732584351, Do you want a New Nokia 3510i c...
                              ...                        
789     5 Free Top Polyphonic Tones call 087018728737,...
968     What do u want when i come back?.a beautiful n...
1667    Guess who spent all last night phasing in and ...
3321    Eh sorry leh... I din c ur msg. Not sad alread...
1688    Free Top ringtone -sub to weekly ringtone-get ...
Name: Message, Length: 4457, dtype: object


In [ ]:
print(X_train_features)

  (0, 2329)	0.38783870336935383
  (0, 3811)	0.34780165336891333
  (0, 2224)	0.413103377943378
  (0, 4456)	0.4168658090846482
  (0, 5413)	0.6198254967574347
  (1, 3811)	0.17419952275504033
  (1, 3046)	0.2503712792613518
  (1, 1991)	0.33036995955537024
  (1, 2956)	0.33036995955537024
  (1, 2758)	0.3226407885943799
  (1, 1839)	0.2784903590561455
  (1, 918)	0.22871581159877646
  (1, 2746)	0.3398297002864083
  (1, 2957)	0.3398297002864083
  (1, 3325)	0.31610586766078863
  (1, 3185)	0.29694482957694585
  (1, 4080)	0.18880584110891163
  (2, 6601)	0.6056811524587518
  (2, 2404)	0.45287711070606745
  (2, 3156)	0.4107239318312698
  (2, 407)	0.509272536051008
  (3, 7414)	0.8100020912469564
  (3, 2870)	0.5864269879324768
  (4, 2870)	0.41872147309323743
  (4, 487)	0.2899118421746198
  :	:
  (4454, 2855)	0.47210665083641806
  (4454, 2246)	0.47210665083641806
  (4455, 4456)	0.24920025316220423
  (4455, 3922)	0.31287563163368587
  (4455, 6916)	0.19636985317119715
  (4455, 4715)	0.30714144758811196
  (

Training the machine learning model

LOGISTIC REGRESSION

In [ ]:
model = LogisticRegression()

In [ ]:
#training the logistic reg model with the training data
model.fit(X_train_features , Y_train)

LogisticRegression()

Evaluvating the trained model.
give x_train to the model and ask for y_train

In [ ]:
#prediction on training data
prediction_on_training_data = model.predict(X_train_features)
accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)

In [ ]:
print('accuracy on training data : ' , accuracy_on_training_data)

accuracy on training data :  0.9676912721561588


accuracy score > 95% is a very good score

Incase of overfit we need to check accuracy for both training and testing data

In [ ]:
#prediction on test data
prediction_on_test_data = model.predict(X_test_features)
accuracy_on_test_data = accuracy_score(Y_test, prediction_on_test_data)

In [ ]:
print('accuracy on test data : ' , accuracy_on_test_data)

accuracy on test data :  0.9668161434977578


Build a predictive system

In [ ]:
input_mail = ["AFree entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's"]

#convert text to feature vector
input_data_features = feature_extraction.transform(input_mail)

#making prediction
prediction = model.predict(input_data_features)
print(prediction)

if prediction [0] == 1 :
  print('Ham mail')
else: print('Spam mail')

[0]
Spam mail
